### PR TITLE
Emit build operation for executing taskgraph whenReady hooks

### PR DESCRIPTION
**Pull request by breskeby**
### Context
This will assist build scans to separate out work done during these hooks.

### Contributor Checklist
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x ] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java [hunk 1]
```diff
@@ -289,4 +294,31 @@ public void execute(final TaskInternal task) {
     public TaskExecutionListener getTaskExecutionListenerSource() {
         return taskListeners.getSource();
     }
+
+    private static class NotifyTaskGraphWhenReady implements RunnableBuildOperation {
+
+        private DefaultTaskExecutionGraph taskExecutionGraph;
```
Comment (ldaley) [inline subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java]: `final`

File: subprojects/core/src/integTest/groovy/org/gradle/configuration/project/LifecycleProjectEvaluatorIntegrationTest.groovy [hunk 2]
```diff
@@ -151,6 +156,11 @@ class LifecycleProjectEvaluatorIntegrationTest extends AbstractIntegrationSpec {
             children*.displayName == ["Apply script after.gradle to project ':foo'"]
             parentId == configOp.id
         }
+        with(operations.only(NotifyTaskGraphWhenReadyBuildOperationType, { it.details.buildPath == ':' })) {
```
Comment (ldaley) [inline subprojects/core/src/integTest/groovy/org/gradle/configuration/project/LifecycleProjectEvaluatorIntegrationTest.groovy]: Would be worth some composite build coverage, and buildSrc.

Comment (breskeby) [reply subprojects/core/src/integTest/groovy/org/gradle/configuration/project/LifecycleProjectEvaluatorIntegrationTest.groovy]: added. Interestingly (to keep in mind) the taskgraph when ready evaluation of the included build is optional, depending on if any task of the included build is triggered or not

File: subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java (hunk 1 above)
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java]: fixed

File: subprojects/core/src/integTest/groovy/org/gradle/configuration/project/LifecycleProjectEvaluatorIntegrationTest.groovy [hunk 3]
```diff
@@ -152,6 +172,24 @@ class LifecycleProjectEvaluatorIntegrationTest extends AbstractIntegrationSpec {
             parentId == configOp.id
         }
 
+        with(operations.only(NotifyTaskGraphWhenReadyBuildOperationType, { it.details.buildPath == ':buildSrc' })) {
+            displayName == 'Notify taskgraph whenReady listeners (:buildSrc)'
```
Comment (ldaley) [inline subprojects/core/src/integTest/groovy/org/gradle/configuration/project/LifecycleProjectEvaluatorIntegrationTest.groovy]: I missed a small thing the first time. Should be `task graph`, not `taskgraph`.

**Pull request merged**